marklogic-data-hub/src/main/java/com/marklogic/hub/impl/ArtifactManagerImpl.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2012-2020 MarkLogic Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.marklogic.hub.impl;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.marklogic.client.DatabaseClient;
+import com.marklogic.hub.ArtifactManager;
+import com.marklogic.hub.HubConfig;
+import com.marklogic.hub.dataservices.ArtifactService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Iterator;
+
+@Component
+public class ArtifactManagerImpl implements ArtifactManager {
+    @Autowired
+    HubConfig hubConfig;
```

Comment:
I'd really like to get away from anything depending on HubConfig to be a singleton. I took a look at the code, and I think it'd be better for ArtifactManagerImpl not to be a Spring component. Rather, just instantiate one when needed by passing in a HubProject and a DatabaseClient (which would be constructed based on a HubConfig, which the client presumably has access to). 

I looked at how allArtifactTypeInfo is used, because if this is no longer a singleton, there won't be a reason to store that as a class field. I think that's fine though. The data is needed during updateArtifact and deleteArtifact. Those should be very fast operations, such that I think it's fine to make a call to ML each time to get the needed artifact type info (particularly since that endpoint looks like it'll be really fast).

Reply:
I think we can set member allArtifactTypeInfo as static, and some methods can be public static methods. My initial thought is why we have to get ArtifactTypeInfo from ML server side if we already know those pre-defined supported artifact types, file extensions ? They can simply set as a final static Map without having to keep the allArtifactTypeInfo object instance with extra Json operations.

Reply:
If we create different objects for the Singleton (ArtifactManagerImpl), it can cause unnecessary memory footprint...

Reply:
The memory footprint of ArtifactManagerImpl would be as minimal as possible though, as it would have no state in it. And they'd be eligible for garbage collection immediately after a controller instantiates one and uses it. We'd be better off worrying about creating instances of String. 

If we feel it's important for ArtifactManagerImpl to be a singleton, then I recommend moving HubConfig from being a class field to being a method argument instead. That's my main goal - that we shouldn't have any classes autowiring in a HubConfig object since that object is tied to one user, and we know that multiple users need to be able to use an object like ArtifactManagerImpl. 

Note that since we don't have a use case for multiple HubProject instances, I think it's fine for that to be a singleton that Spring manages. So having ArtifactManagerImpl autowire in HubProject is fine.

Reply:
Also, I think we want the ML endpoint to be the one that knows about artifact type info. That's knowledge that's derived from code in ML. 

I just wouldn't worry about caching this stuff in ML yet, I think it's premature optimization. Artifacts aren't going to be updated/deleted that often - maybe dozens of times a day? Even hundreds a day would be negligible from a performance perspective. This just isn't something we need to optimize yet, or probably ever.